crates/sui-bridge-indexer/src/server.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::storage::PgBridgePersistent;
+use axum::{
+    debug_handler,
+    extract::{Query, State},
+    http::StatusCode,
+    routing::get,
+    Json, Router,
+};
+use diesel::sql_query;
+use diesel::sql_types::{Bool, Bytea, Int4, Int8, Text};
+use diesel::{prelude::*, sql_types::Nullable};
+use diesel_async::RunQueryDsl;
+use hex;
+use serde::ser::Serializer;
+use serde::{Deserialize, Serialize};
+use std::net::SocketAddr;
+use tokio::{net::TcpListener, task::JoinHandle};
+
+pub const GET_TOKEN_TRANSFERS: &str = "/token_transfers";
+
+pub fn run_server(socket_address: SocketAddr, state: PgBridgePersistent) -> JoinHandle<()> {
+    tokio::spawn(async move {
+        let listener = TcpListener::bind(socket_address).await.unwrap();
+        axum::serve(listener, make_router(state)).await.unwrap();
+    })
+}
+
+pub(crate) fn make_router(state: PgBridgePersistent) -> Router {
+    Router::new()
+        .route("/", get(health_check))
+        .route(GET_TOKEN_TRANSFERS, get(get_token_transfers))
+        .with_state(state)
+}
+
+impl axum::response::IntoResponse for BridgeIndexerError {
+    // TODO: distinguish client error.
+    fn into_response(self) -> axum::response::Response {
+        match self {
+            BridgeIndexerError::InternalError(message) => (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                format!("Internal server error: {}", message),
+            )
+                .into_response(),
+            BridgeIndexerError::InvalidInput(message) => (
+                StatusCode::BAD_REQUEST,
+                format!("Invalid input: {}", message),
+            )
+                .into_response(),
+        }
+    }
+}
+
+async fn health_check() -> StatusCode {
+    StatusCode::OK
+}
+
+#[derive(Deserialize)]
+struct QueryParams {
+    page: usize,
+    per_page: usize,
+    chain_id: i32,
+    address: Option<String>,
+}
+
+#[debug_handler]
+async fn get_token_transfers(
+    params: Query<QueryParams>,
+    State(state): State<PgBridgePersistent>,
+) -> Result<Json<Vec<TokenTransferResult>>, BridgeIndexerError> {
+    let connection = &mut state.pool.get().await?;
+
+    // Destructure params
+    let QueryParams {
+        page,
+        per_page,
+        chain_id,
+        address,
+    } = params.0;
+
+    let offset = (page * per_page) as i64;
+    let limit = per_page as i64;
+
+    let sql = r#"
+        SELECT DISTINCT ON (tt.chain_id, tt.nonce)
```

Comment:
what does this `DISTINCT ON` do along with
```
            CASE tt.status
                WHEN 'Claimed' THEN 3
                WHEN 'Approved' THEN 2
                WHEN 'Deposited' THEN 1
                ELSE 0
            END DESC
```
?

Reply:
The DISTINCT ON clause selects only one row for each unique chain_id and nonce combination that is determined by the ORDER BY clause. The ORDER BY clause prioritises rows by nonce in descending order and then ranks the tt.status values (Claimed, Approved, Deposited) so that the row with the most "recent" status is selected. 

TLDR: this ensures that for a given "token transfer", only one record is returned (instead of at most 3)